Fix data labels too dense with large single series

JesseSDevaney opened the pull request:
- Closes: https://github.com/metabase/metabase/issues/40196
- Closes: https://github.com/metabase/metabase/issues/39568

**Demo** _(updated)_

https://github.com/metabase/metabase/assets/22608765/c808cbf9-bebc-43f3-a3be-99f7f083729a

JesseSDevaney committed:
remove unused variable

JesseSDevaney committed:
conditionally render labels sparsely if large amount of data points

JesseSDevaney committed:
update loki snapshots

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

alxnddr reviewed (changes requested):
I think we need to tweak the logic based on chart width, data points density, and maybe data labels widths becuase it looks too rare:
<img width="870" alt="Screenshot 2024-05-22 at 3 06 11 PM" src="https://github.com/metabase/metabase/assets/14301985/bf477c85-9534-4a94-8aa4-5dbe7813aebf">

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
refactor data label formatter conditions

JesseSDevaney committed:
setup intelligent label plotting

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
handle the case when there are no labels

JesseSDevaney committed:
add sparse data label rendering to bar charts

JesseSDevaney committed:
do not compute if no labels need to be shown or hidden

JesseSDevaney committed:
add sparse data labeling to waterfall charts

JesseSDevaney committed:
adjustment for large series waterfall labels

JesseSDevaney committed:
revert change

JesseSDevaney committed:
refactor name

JesseSDevaney committed:
add sparse labels to stacked bar and area

JesseSDevaney committed:
improve scaleFactor

JesseSDevaney committed:
fix type errors

JesseSDevaney committed:
remove unused labelFormatter

cdeweyx reviewed (approved):
LGTM other than Sasha's comments re rare labels

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
update snapshots

JesseSDevaney committed:
patch ECharts

- labels with no text, i.e. "", were still being considered as plot-able labels which was causing labels with real text "asdfec.." to be hidden because of overlap.
- This patch makes it so that labels with no text, i.e. "", are not considered to be candidates for overlap checking.

JesseSDevaney committed:
update loki snapshots

JesseSDevaney committed:
only loop over dataset once for getWaterfallChartDataDensity

JesseSDevaney committed:
improve performance of chart data density calculations

JesseSDevaney committed:
update loki snapshots

JesseSDevaney committed:
update E2E spec

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
update E2E spec

alxnddr reviewed (commented):
I'd slightly tweak the logic to make it more dense so cases like this show all labels since there is enough space for this:
<img width="884" alt="Screenshot 2024-05-30 at 11 19 44 AM" src="https://github.com/metabase/metabase/assets/14301985/3944e726-ee25-4134-bb23-0dbb0c2c37e8">

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
increase cartesian label density allotment

JesseSDevaney committed:
update loki snapshots

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
[ci nocache] update loki snapshots

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
comment out node_modules cache since it cannot be skipped by commit and is breaking CI

JesseSDevaney committed:
fix type errors

uladzimirdev committed:
Reset node modules if patches are changed

uladzimirdev committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/model/series.ts:
```diff
@@ -324,6 +330,225 @@ function shouldRenderCompact(
+
+    // if we will not be displaying any labels, we do not have to calculate the
+    // label statistics
+    if (
```
Seems like we don't need to iterate through the data at all if `!settings["graph.show_values"] ||
      settings["graph.label_value_frequency"] === "all"` so we can move it out of the forEach

JesseSDevaney replied:
We still will need to determine how many total dots there will be, and we use that to automatically hide them if there are a lot of them. 

That conditional is only preventing label statistics from being computed since those settings relate to labels.

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/model/types.ts:
```diff
@@ -192,7 +192,28 @@ export type StackModel = {
+};
+
+export type CartesianChartDataDensity = BaseChartDataDensity & {
+  type: "cartesian";
```
nit: "combo" is a better name since waterfall is a cartesian chart

alxnddr reviewed (approved):
Looks good, I left two minor comments, please have a look before merging

alxnddr commented on patches/echarts+5.5.0.patch:
```diff
@@ -69,3 +70,21 @@ index c25de1b..f94690e 100644
+index a75f606..57bec85 100644
+--- a/node_modules/echarts/lib/label/labelLayoutHelper.js
++++ b/node_modules/echarts/lib/label/labelLayoutHelper.js
+@@ -256,6 +256,13 @@ export function hideOverlap(labelList) {
```
It would be possible to get this behavior without patching ECharts by using labelLayout and shifting empty labels to one single place outside of the chart but it would be quite hackish. So I agree that ECharts should ignore empty labels when computing layout

JesseSDevaney committed:
refactor type naming

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series

JesseSDevaney committed:
Merge branch 'master' into fix-40196/data-labels-too-dense-with-large-single-series
